consensus/istanbul/validator/weighted.go

```diff
@@ -586,41 +586,89 @@ func (valSet *weightedCouncil) Refresh(hash common.Hash, blockNum uint64, stakin
+	logger.Debug("New proposers calculated", "new proposers", valSet.proposers)
+
+	return nil
+}
+
+// getStakingAmountsOfValidators calculates stakingAmounts of validators.
+// If validators have multiple staking contracts, stakingAmounts will be a sum of stakingAmounts with the same rewardAddress.
+//  - []*weightedValidator : a list of validators which type is converted to weightedValidator
+//  - []float64 : a list of stakingAmounts.
+func (valSet *weightedCouncil) getStakingAmountsOfValidators(stakingInfo *reward.StakingInfo) ([]*weightedValidator, []float64, error) {
+	numValidators := len(valSet.validators)
 	weightedValidators := make([]*weightedValidator, numValidators)
 	stakingAmounts := make([]float64, numValidators)
-	totalStaking := float64(0)
+	addedStaking := make([]bool, len(stakingInfo.CouncilNodeAddrs))
+
 	for vIdx, val := range valSet.validators {
 		weightedVal, ok := val.(*weightedValidator)
 		if !ok {
-			return errors.New(fmt.Sprintf("not weightedValidator. val=%s", val.Address().String()))
+			return nil, nil, errors.New(fmt.Sprintf("not weightedValidator. val=%s", val.Address().String()))
 		}
 		weightedValidators[vIdx] = weightedVal
 
-		i, err := newStakingInfo.GetIndexByNodeId(weightedVal.address)
+		sIdx, err := stakingInfo.GetIndexByNodeAddress(weightedVal.address)
 		if err == nil {
-			weightedVal.SetRewardAddress(newStakingInfo.CouncilRewardAddrs[i])
-			tempStakingAmount := float64(newStakingInfo.CouncilStakingAmounts[i])
-			if newStakingInfo.UseGini {
-				tempStakingAmount = math.Round(math.Pow(tempStakingAmount, 1.0/(1+newStakingInfo.Gini)))
-			}
-			stakingAmounts[vIdx] = tempStakingAmount
-			totalStaking += tempStakingAmount
+			rewardAddr := stakingInfo.CouncilRewardAddrs[sIdx]
+			weightedVal.SetRewardAddress(rewardAddr)
+			stakingAmounts[vIdx] = float64(stakingInfo.CouncilStakingAmounts[sIdx])
+			addedStaking[sIdx] = true
 		} else {
 			weightedVal.SetRewardAddress(common.Address{})
 		}
 	}
 
-	// Update each validator's weight based on the ratio of its staking amount vs. the total staking amount.
+	for sIdx, isAdded := range addedStaking {
+		if isAdded {
+			continue
+		}
+		for vIdx, val := range weightedValidators {
+			if val.RewardAddress() == stakingInfo.CouncilRewardAddrs[sIdx] {
+				stakingAmounts[vIdx] += float64(stakingInfo.CouncilStakingAmounts[sIdx])
+				break
+			}
+		}
+	}
+
+	logger.Debug("stakingAmounts of validators", "validators", weightedValidators, "stakingAmounts", stakingAmounts)
+	return weightedValidators, stakingAmounts, nil
+}
+
+// calcTotalAmount calculates totalAmount of stakingAmounts
+// If UseGini is true, gini is reflected to stakingAmounts
+func calcTotalAmount(stakingInfo *reward.StakingInfo, stakingAmounts []float64) float64 {
+	totalStaking := float64(0)
+	if stakingInfo.UseGini && len(stakingInfo.CouncilNodeAddrs) != 0 {
```

Comment:
As we dicussed f2f, let's check `len(stakingInfo.CouncilNodeAddrs) != 0 ` first!

Reply:
reflect!!!